Draft: Diving into JTAG (Part 2) (#402)
memfault/interrupt: 2023-11-29-diving-into-jtag-part-2.md

---

> title: Diving into JTAG protocol. Part 2 — Debugging.

Reviewer: Let's match with the first article's title:

```suggestion
title: Diving into JTAG - Debugging (Part 2).
```

Author: 100% agreed - really appreciate the extra eyes!

description:
    "This article explores debugging microcontrollers using JTAG, with a special focus on the ARM Cortex-M architecture and the STM32F407VG microcontroller."

> author: Aliaksandr Kavalchuk

Reviewer: This should render a link back to the author profile!

```suggestion
author: zamuhrishka
```

> tags: arm, mcu, debugging, debugger

Reviewer: Nice! just need to wrap in square brackets (and let's add the cortex-m tag while we're at it!)

```suggestion
tags: [arm, cortex-m, mcu, debugging, debugger]
```

Author: Done! thanks for the tip, now I know for the future :)

---

<!-- excerpt start -->

Part 2 of a JTAG deep-dive series gives an in-depth look at interacting with a microcontroller's memory and engaging with the processor core and debug registers.

<!-- excerpt end -->

{% include newsletter.html %}

{% include toc.html %}

> ![image](https://github.com/memfault/interrupt/assets/33341738/8993d42a-f524-48b5-97bc-83f6c6f89db3)

Reviewer: We'll want to put this images under the `img/` folder in a folder called `jtag-part2` or something similar so we can keep track of them and update them more easily. We can then refer to them with `![](/img/jtag-part2/jtag-image.png)` -- naming them with something descriptive though. We can start with that, but we may need to adjust the image size based on how they render. We can do that like we have [here](https://github.com/memfault/interrupt/blob/master/_posts/2023-11-29-diving-into-jtag-part1.md?plain=1#L53-L55).

Reviewer: I actually don't know how to create a folder in the GitHub UI 🤔 It is easiest to do locally. I can fix that up in the publish PR, no worries

> As noted in my previous article [Diving into JTAG protocol. Part 1 — Overview](https://medium.com/@aliaksandr.kavalchuk/diving-into-jtag-protocol-part-1-overview-fbdc428d3a16), JTAG was initially developed for testing integrated circuits and printed circuit boards. However, its potential for debugging was realized over time, and now JTAG has become the standard protocol for microcontroller debugging. Many Firmware and Embedded engineers first encountered it in this particular context.

Reviewer: Let's link back to our article https://interrupt.memfault.com/blog/diving-into-jtag-part1

Author: Excellent catch, can't believe I didn't do that! Thanks for the astute review

However, there’s a catch: while the use of JTAG in testing is fairly standardized when it comes to debugging, each processor architecture has its unique nuances. With that in mind, this article will focus on debugging using JTAG on the ARM Cortex-M architecture, specifically with the STM32F407VG microcontroller.

> For successful debugging, it’s not enough to simply know the basics of JTAG. It’s also essential to understand the workings of the ARM Debug Access Port, a key component that allows access to the controller’s internal registers and memory. I’ve detailed this module in the article [A Deep Dive into the ARM Debug Access Port](https://piolabs.com/blog/engineering/diving-into-arm-debug-access-port.html). With that groundwork laid, let’s dive into the world of debugging through JTAG.

Reviewer: > Second link starting at beginning is broken - https://piolabs.com/blog/engineering/diving-into-arm-debug-access-port.html

This link works for me 🤔 can you try it again to be sure?

Author: You're right, not seeing that either - must have been user error. Thanks!

Access to these registers via the DAP module is essentially the same as accessing memory. Instead of providing a memory address, you provide the address of the desired register. For instance, in the case of the `DHCSR` register, the address is `0xE000EDF0`. If you let me, I won't delve into an example with the bit sequence since it would be identical to the example from the chapter on memory interaction.

## Closing
Well, we’ve briefly explored how JTAG is used in microcontroller debugging. This topic is big, and it’s challenging to cover it fully within a single article. In this piece, we focused on invasive debugging, while non-invasive debugging wasn’t even mentioned. However, that wasn’t the goal when writing this article. For those who want to delve deeper into this topic, I recommend studying the links list listed below. As for us, we’ll move on, and in the next article, we’ll tackle the subject of testing printed circuit boards using the JTAG protocol.

Thank you for your time! :)

> Next part: [Diving into JTAG. Part 3 — Boundary Scan](https://medium.com/@aliaksandr.kavalchuk/diving-into-jtag-part-3-boundary-scan-17f9975ecc59)

Reviewer: Let's remove the link to part 3. We can add a link to our version once it is published though! 

Also, let's backlink to the original article here like Blake did in the previous one

```suggestion
>**Note**: This article was originally published by Aliaksandr on his blog. You can find the original article [here](https://medium.com/@aliaksandr.kavalchuk/diving-into-jtag-protocol-part-2-debugging-56a566db3cf8)
```

Author: Removed the "Next Part" section from the file (edited directly) and then committed your other suggestion - hope I'm doing this right... taking screenshots of your original comments just in case! Thanks so much Gilly
